ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -1303,3 +1337,233 @@ bool attribute_is_exportable(struct pkcs11_attribute_head *req_attr,
+	case PKCS11_CKA_DECRYPT:
+	case PKCS11_CKA_SIGN:
+	case PKCS11_CKA_SIGN_RECOVER:
+	case PKCS11_CKA_UNWRAP:
+	/*
+	 * TBD: Revisit if we don't support PKCS11_CKA_PUBLIC_KEY_INFO
+	 * Specification mentions that if this attribute is
+	 * supplied as part of a template for C_CreateObject, C_CopyObject or
+	 * C_SetAttributeValue for a private key, the token MUST verify
+	 * correspondence between the private key data and the public key data
+	 * as supplied in CKA_PUBLIC_KEY_INFO. This needs to be
+	 * taken care of when this object type will be implemented
+	 */
+	case PKCS11_CKA_PUBLIC_KEY_INFO:
+		return true;
+	/* Can't be modified once set to CK_FALSE - 12 in Table 10 */
+	case PKCS11_CKA_EXTRACTABLE:
+		return get_bool(obj->attributes, PKCS11_CKA_EXTRACTABLE);
+	/* Can't be modified once set to CK_TRUE - 11 in Table 10 */
+	case PKCS11_CKA_SENSITIVE:
+		return !get_bool(obj->attributes, PKCS11_CKA_SENSITIVE);
+	case PKCS11_CKA_WRAP_WITH_TRUSTED:
+		return !get_bool(obj->attributes, PKCS11_CKA_WRAP_WITH_TRUSTED);
+	default:
+		return false;
+	}
+}
+
+static bool attribute_is_modifiable(struct pkcs11_session *session,
+				    struct pkcs11_attribute_head *req_attr,
+				    struct pkcs11_object *obj,
+				    enum pkcs11_class_id class,
+				    enum processing_func function)
+{
+	/* Check modifiable attributes common to any object */
+	switch (req_attr->id) {
+	case PKCS11_CKA_LABEL:
+		return true;
+	case PKCS11_CKA_TOKEN:
+	case PKCS11_CKA_PRIVATE:
+	case PKCS11_CKA_MODIFIABLE:
+	case PKCS11_CKA_DESTROYABLE:
+		if (function == PKCS11_FUNCTION_COPY)
+			return true;
+		else
+			return false;
+	case PKCS11_CKA_COPYABLE:
+		/*
+		 * TBD - Specification mentions that if the attribute value
+		 * is false it can't be set to true. Reading this it looks
+		 * like it should be possible to modify this attribute by
+		 * C_SetAttributeValue() but this is not marked as modifiable
+		 * in Table 10 neither it is mentioned that it can be modified
+		 * anywhere else. So currenly assuming we can modify it.
+		 */
+		if (function == PKCS11_FUNCTION_MODIFY)
+			/* Only allow change from CK_TRUE -> CK_FALSE */
+			return get_bool(obj->attributes, PKCS11_CKA_COPYABLE);
+		else
+			return false;
+	default:
+		break;
+	}
+
+	/* Attribute checking based on class type */
+	switch (class) {
+	case PKCS11_CKO_SECRET_KEY:
+	case PKCS11_CKO_PUBLIC_KEY:
+	case PKCS11_CKO_PRIVATE_KEY:
+		if (attr_is_modifiable_any_key(req_attr))
+			return true;
+		if (class == PKCS11_CKO_SECRET_KEY) {
+			if (attr_is_modifiable_secret_key(req_attr, session,
+							  obj))
+				return true;
+		}
+		if (class == PKCS11_CKO_PUBLIC_KEY) {
+			if (attr_is_modifiable_public_key(req_attr, session,
+							  obj))
+				return true;
+		}
+		if (class == PKCS11_CKO_PRIVATE_KEY) {
+			if (attr_is_modifiable_private_key(req_attr, session,
+							   obj))
+				return true;
+		}
+		break;
+	case PKCS11_CKO_DATA:
+		/* None of the data object attributes are modifiable */
```

Comment:
I suppose this is not true?

Reply:
As per what I interpret from specification, this is true. I have considered following criteria for deciding if an attribute is modifiable :
1. Marked with 8,10,11 or 12 as per Table 10 in specification
2. Explicitly mentioned as modifiable in the specification

Reply:
OK. Reading the text again and even checked 3.0 specification too. There is no such markings that data objects could be modified.